.github/workflows/links.yml

```diff
@@ -0,0 +1,77 @@
+name: Check Links In Pull Requests
+
+on:
+  pull_request:
+    branches:
+      - main
+    # Limit the check to certain file types
+    paths:
+      - '**/*.md'
+      - '**/*.ipynb'
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  check-links:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Clone repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          ref: ${{github.event.pull_request.head.ref}}
+          repository: ${{github.event.pull_request.head.repo.full_name}}
+
+      - name: Check out base branch
+        run: git checkout ${{github.event.pull_request.base.ref}}
+
+      - name: Dump all links from ${{github.event.pull_request.base.ref}}
+        uses: lycheeverse/lychee-action@v2
+        with:
+          args: |
+            --dump
+            --include-fragments
+            './**/*.ipynb' './**/*.md'
+          output: ./existing-links.txt
+        continue-on-error: true # Don't fail if base branch check has issues
+
+      - name: Stash untracked files
+        run: git stash push --include-untracked
+
+      - name: Check out feature branch
+        run: git checkout ${{ github.head_ref }}
+
+      - name: Apply stashed changes
+        run: git stash pop || true
```

Comment:
If I'm correct all those steps are done to find the links in the base branch and exclude them from the check, right ? If it's correct, could you gather them in fewer steps and add comments to detail the process ? It took me a bit of (my limited) brain power to understand that (especially the stash part).

I think merging the last 3 should cause no problem but it may not be possible to do only one (due to the use of the `lychee-action` in the middle).

Also adding a reference to https://lychee.cli.rs/github_action_recipes/pull-requests/ in the code might be useful (but no need to check this one with `lychee` :).

Reply:
Hi Mathieu, I only copy-pasted the recipe but I don't understand all the steps either. If you want to modify the code, you can do so at your own risk, I think you may have push rights, or merge this one and edit as you see fit.

This discussion should have more info, but frankly I don't think it's worth my time to become a lychee expert

https://github.com/lycheeverse/lychee-action/issues/238

Reply:
I've been thinking about this a bit more. At it is now, the test makes a full pass on the `main` branch and then only seeks for broken links that are newly added on the incoming feature branch. That's why the test is so involved.

I now think that it would be more useful if lychee is run on the entire feature branch, ignoring if the link is newly introduced or not. After all, we **do** want to know about old links that stopped working.

Do you want me to try to modify the test code so that lychee makes a full pass on the entire repo on each PR instead?

Reply:
Hello,

Sure, we won't become experts on lychee. The discussion you pointed out is interesting as it raises some issues with the recipe.

While I was writing this remark, I saw yours.

I'm not sure what to do now. Maybe we can merge this PR and open new issue to discuss improvements.

Reply:
Hello,

Rethinking about this, I think it's better to go for the simpler version (i.e. checking all links) as you suggested.

The rationale is the following:

1. The recipe will be simpler (we avoid the potential problems mentioned in the discussion you pointed out) and more in line with other recipes that check all notebooks whether they have been modified or not.
2. We probably won't check a lot of links (I would say a few dozens) so testing all won't take that long especially since it applies only to `*.ipynb` and `*.md` files (I realize this slightly contradicts the second part of 1).

Reply:
Sure, that sounds reasonable to me. As a side note, I just ran lychee on the notebooks and it's giving me some trouble, so I think I will need to spend more time on this!

Reply:
Hi Mathieu,

I cleaned-up the test code and it now checks for, and reports on, **all** broken links, be it newly introduced or past links.

It's triggered with the same rules as the rest of the tests.

It is already detecting a few dozen broken links and also links that don't use secure `https` protocol. This is captured in the ticket #26.

Reply:
Hello Martin,

Great. I think we can merge this PR now.